drivers/sony/displays/bravia.cr

```diff
@@ -0,0 +1,226 @@
+  def mute?
+    query(:mute, priority: 0)
+  end
+
+  def mute_audio(state : Bool = true)
+    request(:audio_mute, state)
+    audio_mute?
+  end
+
+  def unmute_audio
+    mute_audio false
+  end
+
+  def audio_mute?
+    query(:audio_mute, priority: 0)
+  end
+
+  def volume(level : Int32)
+    request(:volume, level.to_i)
+    volume?
+  end
+
+  def volume?
+    query(:volume, priority: 0)
+  end
+
+  def do_poll
+    if self[:power]?
+      input?
+      mute?
+      audio_mute?
+      volume?
+    end
+  end
+
+  enum MessageType : UInt8
+    Answer  = 0x41
+    Control = 0x43
+    Enquiry = 0x45
+    Notify  = 0x4e
+
+    def control_character
+      case self
+      in Answer  then "A"
+      in Control then "C"
+      in Enquiry then "E"
+      in Notify  then "N"
+      end
+    end
+  end
+
+  def received(data, task)
+    parsed_data = convert_binary(data[3..6])
+    cmd = RESPONSES[parsed_data]
+    param = data[7..-1]
+
+    return task.try(&.abort("error")) if param.first? == ERROR
+    case MessageType.from_value?(data[2])
+    when MessageType::Answer
+      update_status cmd, param
+      task.try &.success
+    when MessageType::Notify
+      update_status cmd, param
+    else
+      logger.debug { "Unhandled device response: #{data[2]}" }
+      task.try &.abort("Unhandled device response")
+    end
+  end
+
+  COMMANDS = {
+    ir_code:           "IRCC",
+    power:             "POWR",
+    volume:            "VOLU",
+    audio_mute:        "AMUT",
+    mute:              "PMUT",
+    channel:           "CHNN",
+    tv_input:          "ISRC",
+    input:             "INPT",
+    toggle_mute:       "TPMU",
+    pip:               "PIPI",
+    toggle_pip:        "TPIP",
+    position_pip:      "TPPP",
+    broadcast_address: "BADR",
+    mac_address:       "MADR",
+  }
+  RESPONSES = COMMANDS.to_h.invert
+
+  protected def convert_binary(data)
+    data.join &.chr
+  end
+
+  protected def request(command, parameter, **options)
+    cmd = COMMANDS[command]
+    parameter = parameter ? 1 : 0 if parameter.is_a?(Bool)
+    param = parameter.to_s.rjust(16, '0')
+    do_send(MessageType::Control, cmd, param, **options)
+  end
+
+  protected def query(state, **options)
+    cmd = COMMANDS[state]
+    do_send(MessageType::Enquiry, cmd, HASH, **options)
+  end
+
+  protected def do_send(type, command, parameter, **options)
+    cmd = "#{INDICATOR}#{type.control_character}#{command}#{parameter}\n"
+    send(cmd, **options)
+  end
+
+  protected def update_status(cmd, param)
+    parsed_data = convert_binary(param)
+    case cmd
+    when :power, :mute, :audio_mute, :pip
+      self[cmd] = parsed_data.to_i == 1
+    when :volume
+      self[:volume] = parsed_data.to_i
+    when :mac_address
+      self[:mac_address] = parsed_data.split('#')[0]
+    when :input
+      input_num = convert_binary(param[7..10])
+      index_num = convert_binary(param[11..-1]).to_i
+      if index_num == 1
+        self[:input] = CONVERT_LOOKUP[input_num]
```

Comment:
Does the `input_num` vary here for members of the input types?
Since this is related to `Input`, it would be neater if the logic was held inside the `Input` enum.